File: src/history.js
// Mercury Code - Session History Manager
// Saves and restores complete conversation sessions for backup and time-travel

import { readFile, writeFile, readdir, mkdir, rm, rename, unlink } from "node:fs/promises";
import path from "node:path";
import { homedir } from "node:os";
import { debugLog } from "./utils/debug-log.js";

const SESSIONS_DIR = path.join(homedir(), ".mercury", "sessions");
const MAX_SESSIONS = 50; // keep at most 50 sessions

/**
 * Redact common secret patterns from session summary text.
 * Prevents API keys, tokens, passwords, and other credentials from
 * being exposed in the /history list.
 * @param {string} text - Raw summary text
 * @returns {string} Sanitized text with secrets redacted
 */
function sanitizeSummary(text) {
  return text
    .replace(/(?:sk|pk|api|key|token|secret|password|bearer)[-_]?[a-zA-Z0-9]{16,}/gi, '[REDACTED]')
    .replace(/(?:eyJ)[A-Za-z0-9+/=]{20,}/g, '[JWT-REDACTED]')
    .replace(/(?:ghp|gho|ghu|ghs|ghr)_[A-Za-z0-9]{36,}/g, '[GITHUB-TOKEN-REDACTED]')
    .replace(/(?:xox[bpsa])-[A-Za-z0-9-]{10,}/g, '[SLACK-TOKEN-REDACTED]');
}

/**
 * SessionHistory manages saved conversation sessions.
 * Each session is stored as a JSON file in ~/.mercury/sessions/
 */
export class SessionHistory {
  constructor(options = {}) {
    this.dir = options.dir || SESSIONS_DIR;
  }

  /**
   * Ensure the sessions directory exists.
   */
  async _ensureDir() {
    await mkdir(this.dir, { recursive: true });
  }

  /**
   * Save the current session to a file.
   * @param {object} session
   * @param {string} session.id - Unique session ID
   * @param {string} session.cwd - Working directory
   * @param {Array} session.messages - Conversation messages
   * @param {object} session.config - Model config snapshot
   * @param {string} [session.summary] - Optional session summary
   * @returns {string} Path to saved session file
   */
  async save(session) {
    await this._ensureDir();
    const ts = new Date().toISOString().replace(/[:.]/g, "-");
    const filename = `${ts}_${session.id}.json`;
    const filepath = path.join(this.dir, filename);

    const data = {
      id: session.id,
      cwd: session.cwd,
      timestamp: Date.now(),
      date: new Date().toISOString(),
      messageCount: session.messages.length,
      summary: session.summary || this._autoSummary(session.messages),
      messages: session.messages,
      config: session.config || {},
      trustMode: session.trustMode || null,
      planMode: session.planMode || false,
      sandbox: session.sandbox || null,
    };

    const tmpPath = filepath + '.tmp-' + Date.now();
    try {
      await writeFile(tmpPath, JSON.stringify(data, null, 2), "utf-8");
      await rename(tmpPath, filepath);
    } catch (err) {
      try { await unlink(tmpPath); } catch {}
      throw err;
    }
    await this._pruneOld();
    return filepath;
  }

  /**
   * Auto-generate a short summary from the first user message.
   */
  _autoSummary(messages) {
    const firstUser = messages.find((m) => m.role === "user");
    if (!firstUser || !firstUser.content) return "(empty session)";
    const text = firstUser.content.slice(0, 80);
    const raw = text.length < firstUser.content.length ? text + "..." : text;
    return sanitizeSummary(raw);
  }

  /**
   * List all saved sessions, sorted by most recent first.
   * @returns {Array<{id, date, messageCount, summary, filename}>}
   */
  async list() {
    await this._ensureDir();
    let files;
    try {
      files = await readdir(this.dir);
    } catch (err) {
      debugLog("SessionHistory.list.readdir", err);
      return [];
    }

    const sessions = [];
    for (const f of files) {
      if (!f.endsWith(".json")) continue;
      try {
        const raw = await readFile(path.join(this.dir, f), "utf-8");
        const data = JSON.parse(raw);
        sessions.push({
          id: data.id,
          date: data.date,
          timestamp: data.timestamp,
          messageCount: data.messageCount,
          summary: data.summary,
          cwd: data.cwd,
          filename: f,
        });
      } catch (err) {
        debugLog("SessionHistory.list.parseSession", err);
      }
    }

    sessions.sort((a, b) => b.timestamp - a.timestamp);
    return sessions;
  }

  /**
   * Load a session by index (from the list) or by filename.
   * @param {string} identifier - Session index (1-based) or filename
   * @returns {object|null} Full session data
   */
  async load(identifier) {
    const sessions = await this.list();

    let session;
    const idx = parseInt(identifier, 10);
    if (!isNaN(idx) && idx >= 1 && idx <= sessions.length) {
      session = sessions[idx - 1];
    } else {
      session = sessions.find(
        (s) => s.filename === identifier || s.id === identifier
      );
    }

    if (!session) return null;

    let raw;
    try {
      raw = await readFile(
        path.join(this.dir, session.filename),
        "utf-8"
      );
      return JSON.parse(raw);
    } catch (err) {
      debugLog("SessionHistory.load", err);
      return null;
    }
  }

  /**
   * Delete old sessions beyond MAX_SESSIONS.
   */
  async _pruneOld() {
    const sessions = await this.list();
    if (sessions.length <= MAX_SESSIONS) return;

    const toDelete = sessions.slice(MAX_SESSIONS);
    for (const s of toDelete) {
      try {
        await rm(path.join(this.dir, s.filename));
      } catch (err) {
        debugLog("SessionHistory._pruneOld", err);
      }
    }
  }
}

File: rust-port/crates/mercury-cli/src/history.rs
// Mercury Code - Session History Manager
// Saves and restores complete conversation sessions for backup and time-travel.
// Ported from: src/history.js

use chrono::Utc;
use regex::Regex;
use serde::{Deserialize, Serialize};
use std::path::{Path, PathBuf};
use tokio::fs;

use crate::utils::debug_log;

/// Maximum number of sessions to keep on disk.
const MAX_SESSIONS: usize = 50;

/// Redact common secret patterns from session summary text.
/// Prevents API keys, tokens, passwords, and other credentials
/// from being exposed in the /history list.
fn sanitize_summary(text: &str) -> String {
    use once_cell::sync::Lazy;

    static RE_GENERIC: Lazy<Regex> = Lazy::new(|| {
        Regex::new(r"(?i)(?:sk|pk|api|key|token|secret|password|bearer)[-_]?[a-zA-Z0-9]{16,}")
            .unwrap()
    });
    static RE_JWT: Lazy<Regex> =
        Lazy::new(|| Regex::new(r"(?:eyJ)[A-Za-z0-9+/=]{20,}").unwrap());
    static RE_GITHUB: Lazy<Regex> =
        Lazy::new(|| Regex::new(r"(?:ghp|gho|ghu|ghs|ghr)_[A-Za-z0-9]{36,}").unwrap());
    static RE_SLACK: Lazy<Regex> =
        Lazy::new(|| Regex::new(r"(?:xox[bpsa])-[A-Za-z0-9-]{10,}").unwrap());

    let s = RE_GENERIC.replace_all(text, "[REDACTED]");
    let s = RE_JWT.replace_all(&s, "[JWT-REDACTED]");
    let s = RE_GITHUB.replace_all(&s, "[GITHUB-TOKEN-REDACTED]");
    let s = RE_SLACK.replace_all(&s, "[SLACK-TOKEN-REDACTED]");
    s.into_owned()
}

/// Saved session data.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionData {
    pub id: String,
    pub cwd: String,
    pub timestamp: i64,
    pub date: String,
    #[serde(rename = "messageCount")]
    pub message_count: usize,
    pub summary: String,
    pub messages: Vec<serde_json::Value>,
    #[serde(default)]
    pub config: serde_json::Value,
    #[serde(default, rename = "trustMode")]
    pub trust_mode: Option<String>,
    #[serde(default, rename = "planMode")]
    pub plan_mode: bool,
    #[serde(default)]
    pub sandbox: Option<serde_json::Value>,
}

/// Session listing entry (metadata only, no messages).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionEntry {
    pub id: String,
    pub date: String,
    pub timestamp: i64,
    pub message_count: usize,
    pub summary: String,
    pub cwd: String,
    pub filename: String,
}

/// SessionHistory manages saved conversation sessions.
/// Each session is stored as a JSON file in ~/.mercury/sessions/
pub struct SessionHistory {
    dir: PathBuf,
}

impl SessionHistory {
    pub fn new(dir: Option<PathBuf>) -> Self {
        let dir = dir.unwrap_or_else(|| {
            dirs::home_dir()
                .unwrap_or_else(|| PathBuf::from("."))
                .join(".mercury")
                .join("sessions")
        });
        Self { dir }
    }

    /// Ensure the sessions directory exists.
    async fn ensure_dir(&self) -> std::io::Result<()> {
        fs::create_dir_all(&self.dir).await
    }

    /// Save the current session to a file.
    pub async fn save(&self, session: &SessionData) -> anyhow::Result<String> {
        self.ensure_dir().await?;

        let ts = Utc::now().format("%Y-%m-%dT%H-%M-%S").to_string();
        let filename = format!("{}_{}.json", ts, session.id);
        let filepath = self.dir.join(&filename);

        let data = SessionData {
            id: session.id.clone(),
            cwd: session.cwd.clone(),
            timestamp: Utc::now().timestamp_millis(),
            date: Utc::now().to_rfc3339(),
            message_count: session.messages.len(),
            summary: if session.summary.is_empty() {
                self.auto_summary(&session.messages)
            } else {
                session.summary.clone()
            },
            messages: session.messages.clone(),
            config: session.config.clone(),
            trust_mode: session.trust_mode.clone(),
            plan_mode: session.plan_mode,
            sandbox: session.sandbox.clone(),
        };

        let json = serde_json::to_string_pretty(&data)?;

        // Atomic write: write to tmp file, then rename
        let tmp_path = filepath.with_extension(format!("tmp-{}", Utc::now().timestamp_millis()));
        match fs::write(&tmp_path, &json).await {
            Ok(()) => {
                if let Err(e) = fs::rename(&tmp_path, &filepath).await {
                    let _ = fs::remove_file(&tmp_path).await;
                    return Err(e.into());
                }
            }
            Err(e) => {
                let _ = fs::remove_file(&tmp_path).await;
                return Err(e.into());
            }
        }

        self.prune_old().await;
        Ok(filepath.to_string_lossy().to_string())
    }

    /// Auto-generate a short summary from the first user message.
    fn auto_summary(&self, messages: &[serde_json::Value]) -> String {
        let first_user = messages.iter().find(|m| {
            m.get("role")
                .and_then(|r| r.as_str())
                .map_or(false, |r| r == "user")
        });

        match first_user {
            Some(msg) => {
                let content = msg
                    .get("content")
                    .and_then(|c| c.as_str())
                    .unwrap_or("");
                let text: String = content.chars().take(80).collect();
                let raw = if text.len() < content.len() {
                    format!("{}...", text)
                } else {
                    text
                };
                sanitize_summary(&raw)
            }
            None => "(empty session)".to_string(),
        }
    }

    /// List all saved sessions, sorted by most recent first.
    pub async fn list(&self) -> Vec<SessionEntry> {
        if self.ensure_dir().await.is_err() {
            return Vec::new();
        }

        let mut entries = Vec::new();
        let mut read_dir = match fs::read_dir(&self.dir).await {
            Ok(rd) => rd,
            Err(e) => {
                debug_log("SessionHistory.list.read_dir", &e);
                return Vec::new();
            }
        };

        while let Ok(Some(entry)) = read_dir.next_entry().await {
            let filename = entry.file_name().to_string_lossy().to_string();
            if !filename.ends_with(".json") {
                continue;
            }
            match fs::read_to_string(entry.path()).await {
                Ok(raw) => match serde_json::from_str::<SessionData>(&raw) {
                    Ok(data) => {
                        entries.push(SessionEntry {
                            id: data.id,
                            date: data.date,
                            timestamp: data.timestamp,
                            message_count: data.message_count,
                            summary: data.summary,
                            cwd: data.cwd,
                            filename,
                        });
                    }
                    Err(e) => {
                        debug_log("SessionHistory.list.parse", &e);
                    }
                },
                Err(e) => {
                    debug_log("SessionHistory.list.read", &e);
                }
            }
        }

        entries.sort_by(|a, b| b.timestamp.cmp(&a.timestamp));
        entries
    }

    /// Load a session by index (1-based from the list) or by filename/id.
    pub async fn load(&self, identifier: &str) -> Option<SessionData> {
        let sessions = self.list().await;

        let session = if let Ok(idx) = identifier.parse::<usize>() {
            if idx >= 1 && idx <= sessions.len() {
                sessions.get(idx - 1)
            } else {
                None
            }
        } else {
            sessions
                .iter()
                .find(|s| s.filename == identifier || s.id == identifier)
        };

        let session = session?;
        let filepath = self.dir.join(&session.filename);

        match fs::read_to_string(&filepath).await {
            Ok(raw) => match serde_json::from_str::<SessionData>(&raw) {
                Ok(data) => Some(data),
                Err(e) => {
                    debug_log("SessionHistory.load.parse", &e);
                    None
                }
            },
            Err(e) => {
                debug_log("SessionHistory.load.read", &e);
                None
            }
        }
    }

    /// Delete old sessions beyond MAX_SESSIONS.
    async fn prune_old(&self) {
        let sessions = self.list().await;
        if sessions.len() <= MAX_SESSIONS {
            return;
        }

        for s in sessions.iter().skip(MAX_SESSIONS) {
            let path = self.dir.join(&s.filename);
            if let Err(e) = fs::remove_file(&path).await {
                debug_log("SessionHistory.prune_old", &e);
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sanitize_summary_redacts_api_key() {
        let text = "Set key to sk-abc1234567890abcdef";
        let sanitized = sanitize_summary(text);
        assert!(sanitized.contains("[REDACTED]"));
        assert!(!sanitized.contains("sk-abc"));
    }

    #[test]
    fn test_sanitize_summary_redacts_jwt() {
        let text = "token eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9abcdef";
        let sanitized = sanitize_summary(text);
        assert!(sanitized.contains("[JWT-REDACTED]"));
    }

    #[test]
    fn test_sanitize_summary_redacts_github() {
        let text = "ghp_abcdefghijklmnopqrstuvwxyz1234567890";
        let sanitized = sanitize_summary(text);
        assert!(sanitized.contains("[GITHUB-TOKEN-REDACTED]"));
    }

    #[test]
    fn test_sanitize_summary_passthrough() {
        let text = "normal conversation about code";
        let sanitized = sanitize_summary(text);
        assert_eq!(sanitized, text);
    }

    #[tokio::test]
    async fn test_session_history_new() {
        let tmp = tempfile::TempDir::new().unwrap();
        let history = SessionHistory::new(Some(tmp.path().to_path_buf()));
        let sessions = history.list().await;
        assert!(sessions.is_empty());
    }
}
